SignalR C++ Client (#11196)
aspnet/AspNetCore.Docs: aspnetcore/signalr/cpp-client.md

---

> title: ASP.NET Core SignalR C++ Client

Reviewer: ```suggestion
title: ASP.NET Core SignalR C++ client
```

> author: bradygaster

Reviewer 2: everyone's putting my name in here. :) Feel free to use your own so you can get feedback on your docs. Or, if you want me to police them for you throw me under the bus. :)

Author: Think of it as me giving you credit for the C++ client 😛

> description: Information about the ASP.NET Core SignalR C++ Client

Reviewer: ```suggestion
description: Information about the ASP.NET Core SignalR C++ client
```

monikerRange: '>= aspnetcore-3.0'
ms.author: bradyg
ms.custom: mvc

> ms.date: 2/26/2019

Reviewer: ```suggestion
ms.date: 02/26/2019
```

uid: signalr/cpp-client
---

> # ASP.NET Core SignalR C++ Client

Reviewer: ```suggestion
# ASP.NET Core SignalR C++ client
```

> The ASP.NET Core SignalR C++ client library lets you communicate with SignalR hubs from native applications.

Reviewer 2: Might we add some verbiage here about what benefits that'd offer? Like "why would you care," as a reader?

Author: Idk, I feel like if you're reading a doc on a C++ client, you should already be aware of the usefulness to you. Otherwise you wouldn't have clicked the link.

## Consume the SignalR C++ client library

TODO

## Connect to a hub

```c++
signalr::hub_connection connection("http://localhost:5000/default", signalr::trace_level::all);
connection.start().get();
```

## Handle lost connection

Register a handler for the disconnected event to respond to a lost connection. For example, you might want to automate reconnection.

> The disconnected event can be registered by calling `set_disconnected` and passing in a function;

Reviewer: ```suggestion
The disconnected event can be registered by calling `set_disconnected` and passing in a function:
```

Author: Haha! Classic dev mistake;

```c++
connection.set_disconnected([]()
{
    // Do your close logic.
});
```

## Call hub methods from client

> `connection.send` calls methods on the hub and does not expect or wait for a response.

Reviewer: ```suggestion
`connection.send` calls methods on the hub and doesn't expect or wait for a response.

```

```c++
web::json::value args{};
args[0] = web::json::value::string("some text");
connection.send("Echo", args).get();
```

> `connection.invoke` calls methods on the hub and waits for a response.

Reviewer: ```suggestion
`connection.invoke` calls methods on the hub and waits for a response.

```

```c++
web::json::value args{};
args[0] = web::json::value::string("some text");
connection.invoke("Echo", args)
    .then([](const web::json::value& value)
    {
        ucout << value.serialize() << std::endl;
    }).get();
```

Pass the hub method name and any arguments defined in the hub method to `invoke` or `send`.

## Call client methods from hub

> Define methods the hub calls using `connection.on` after creating the connection, but before starting it. The `web::json::value` parameter will be an array of values sent from the server.

Reviewer: ```suggestion
Define methods the hub calls using `connection.on` after creating the connection, but before starting it. The `web::json::value` parameter will be an array of values sent from the server.

```

```c++
connection.on("ReceiveMessage", [](const web::json::value& m)
{
    ucout << m.at(0).as_string() << m.at(1).as_string() std::endl;
});
```

The preceding code in `connection.on` runs when server-side code calls it using the `SendAsync` method.

[!code-csharp[Call client method](dotnet-client/sample/signalrchat/hubs/chathub.cs?name=snippet_SendMessage)]

## Custom logging

To add logging to the client create a class that inherits from `signalr::log_writer` and add a `write` method that overrides the pure virtual base class.

```c++
class custom_logger : public signalr::log_writer
{
    virtual void write(const utility::string_t& log) override
    {
        ucout << log << std::endl;
    }
}
```

> Then when creating the hub_connection pass a `shared_ptr<custom_logger>` into the constructor.

Reviewer: ```suggestion
When creating the `hub_connection`, pass a `shared_ptr<custom_logger>` into the constructor.
```

---
title: ASP.NET Core SignalR C++ client
author: bradygaster
description: Information about the ASP.NET Core SignalR C++ client
monikerRange: '>= aspnetcore-3.0'
ms.author: bradyg
ms.custom: mvc
ms.date: 02/26/2019
uid: signalr/cpp-client
---

# ASP.NET Core SignalR C++ client

The ASP.NET Core SignalR C++ client library lets you communicate with SignalR hubs from native applications.

## Consume the SignalR C++ client library

TODO

## Connect to a hub

> ```c++

Reviewer: ```suggestion
```cpp
```

signalr::hub_connection connection("http://localhost:5000/default", signalr::trace_level::all);
connection.start().get();
```

## Handle lost connection

Register a handler for the disconnected event to respond to a lost connection. For example, you might want to automate reconnection.

The disconnected event can be registered by calling `set_disconnected` and passing in a function:

> ```c++

Reviewer: ```suggestion
```cpp
```

connection.set_disconnected([]()
{
    // Do your close logic.
});
```

## Call hub methods from client

`connection.send` calls methods on the hub and doesn't expect or wait for a response.

> ```c++

Reviewer: ```suggestion
```cpp
```

web::json::value args{};
args[0] = web::json::value::string("some text");
connection.send("Echo", args).get();
```

`connection.invoke` calls methods on the hub and waits for a response.

> ```c++

Reviewer: ```suggestion
```cpp
```

web::json::value args{};
args[0] = web::json::value::string("some text");
connection.invoke("Echo", args)
    .then([](const web::json::value& value)
    {
        ucout << value.serialize() << std::endl;
    }).get();
```

Pass the hub method name and any arguments defined in the hub method to `invoke` or `send`.

> ## Call client methods from hub

Reviewer 3: I'd move this section up, right after the "Connect to a Hub"

> Define methods the hub calls using `connection.on` after creating the connection, but before starting it. The `web::json::value` parameter will be an array of values sent from the server.

Reviewer 3: For "but before starting it" you could note that you could miss messages otherwise

> ```c++

Reviewer: ```suggestion
```cpp
```

connection.on("ReceiveMessage", [](const web::json::value& m)
{
    ucout << m.at(0).as_string() << m.at(1).as_string() std::endl;
});
```

> The preceding code in `connection.on` runs when server-side code calls it using the `SendAsync` method.

Reviewer 3: Maybe something like "The handler registered in our `connection.on` method is triggered when we receive an invocation from the server for the "ReceiveMessage" method"

[!code-csharp[Call client method](dotnet-client/sample/signalrchat/hubs/chathub.cs?name=snippet_SendMessage)]

## Custom logging

To add logging to the client create a class that inherits from `signalr::log_writer` and add a `write` method that overrides the pure virtual base class.

> ```c++

Reviewer: ```suggestion
```cpp
```

class custom_logger : public signalr::log_writer
{
    virtual void write(const utility::string_t& log) override
    {
        ucout << log << std::endl;
    }
}
```

When creating the `hub_connection`, pass a `shared_ptr<custom_logger>` into the constructor.

> ```c++

Reviewer: ```suggestion
```cpp
```

signalr::hub_connection connection("http://localhost:5000/default", signalr::trace_level::all, std::make_shared<custom_logger>());
```

## Additional resources

> * [Hubs](xref:signalr/hubs)

Reviewer 3: Reminds me that you will need to add an entry into the Table of contents under the "Clients" section.
